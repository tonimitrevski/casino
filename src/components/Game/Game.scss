@import "../../style/mixins";
@import "../../style/variables";

.Game {
  flex-basis: 50%;
  padding: 10px;
  &__container {
    cursor: pointer;
    padding: 8px;
    position: relative;
    &__ribbon {
      pointer-events: none;
      position: absolute;
      right: 5px;
      top: 5px;
      z-index: 1;
      overflow: hidden;
      width: 75px;
      height: 75px;
      text-align: right;

      &__label {
        font-size: 10px;
        font-weight: 700;
        color: $color-ribon-label-buttons;
        text-transform: uppercase;
        text-align: center;
        line-height: 20px;
        -webkit-transform: rotate(45deg);
        transform: rotate(45deg);
        width: 100px;
        display: block;
        background: $background-ribon-label-buttons;
        box-shadow: 0 3px 10px -5px #000;
        position: absolute;
        top: 19px;
        right: -21px;
      }
    }
    &__jackpot {
      background-color: rgba(0, 0, 0, 0.3);
      position: absolute;
      width: 93%;
      display: flex;
      justify-content: center;
      top: 15px;
      color: $color-ribon-label-buttons;
      padding: 3px 0;
      span {
        font-weight: bold;
      }
    }
  }
  img {
    width: 100%;
  }
  @include medium {
    flex-basis: 33%;
  }
  @include large {
    flex-basis: 25%;
  }
  @include ex-large {
    flex-basis: 20%;
  }
}
