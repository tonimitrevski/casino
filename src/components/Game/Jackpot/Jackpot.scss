@import "../../../style/variables";

.jackpot {
  background-color: rgba(0, 0, 0, 0.3);
  position: absolute;
  width: 93%;
  display: flex;
  justify-content: center;
  top: 15px;
  color: $color-ribon-label-buttons;
  padding: 3px 0;
  span {
    font-weight: bold;
  }
}
