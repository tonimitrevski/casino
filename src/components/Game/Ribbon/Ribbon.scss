@import "../../../style/variables";

.ribbon {
  pointer-events: none;
  position: absolute;
  right: 5px;
  top: 5px;
  z-index: $z-index-game;
  overflow: hidden;
  width: 75px;
  height: 75px;
  text-align: right;

  &__label {
    font-size: 10px;
    font-weight: 700;
    color: $color-ribon-label-buttons;
    text-transform: uppercase;
    text-align: center;
    line-height: 20px;
    -webkit-transform: rotate(45deg);
    transform: rotate(45deg);
    width: 100px;
    display: block;
    background: $background-active-buttons;
    box-shadow: 0 3px 10px -5px #000;
    position: absolute;
    top: 19px;
    right: -21px;
  }
}
