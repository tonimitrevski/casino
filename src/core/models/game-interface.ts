export interface GameInterface {
    categories: Array<string>;
    id: string;
    image: string;
    name: string;
}
