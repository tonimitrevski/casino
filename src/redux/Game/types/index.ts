import {FetchGamesAction} from "./fetch-games-action.type";

export type GamesActionTypes = FetchGamesAction;
