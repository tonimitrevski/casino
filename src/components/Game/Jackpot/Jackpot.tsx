import "./Jackpot.scss";
import React from "react";

export const Jackpot = (props : { amount: number }) => {
  const { amount } = props;
  return (
    <div className="jackpot">
      <span>{currencyFormat(amount)}</span>
    </div>
  );
}

function currencyFormat(amount: number) {
  return new Intl.NumberFormat('en-UK', {
    style: 'currency',
    currency: 'GBP',
    minimumFractionDigits: 2
  }).format(amount)
}

export default Jackpot;
