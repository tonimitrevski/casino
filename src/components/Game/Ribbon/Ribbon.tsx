import React from "react";
import PropTypes from "prop-types";
import "./Ribbon.scss";

const Ribbon = (props : { type: string }) => {
  const { type } = props;
  return (
    <div className="ribbon">
      <span className="ribbon__label">{type}</span>
    </div>
  );
}

Ribbon.propTypes = {
  type: PropTypes.string,
};

export default Ribbon;
