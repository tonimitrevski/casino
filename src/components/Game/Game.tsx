import React, {useEffect, useState} from "react";
import "./Game.scss";
import {GameInterface} from "../../core/models/game-interface";
import {connect} from "react-redux";
import {StoreType} from "../../redux/store.type";
import {getJackpotsSelector} from "../../redux/Jackpot/selectors";
import {JackpotStateInterface} from "../../redux/Jackpot/types/jackpot-state-interface";
import {JackpotsAggregate} from "../../core/aggregate/jackpots.aggregate";
import Ribbon from "./Ribbon/Ribbon";
import {Jackpot} from "./Jackpot/Jackpot";
import {FetchImageWorker} from "../../core/services/fetch-image.worker";
import LazyLoad from "react-lazyload";
import Placeholder from "../Placeholder/Placeholder";

export const Game = (props: { game: GameInterface, jackpots: JackpotStateInterface }) => {
  const { game, jackpots } = props;
  const [image, setImage] = useState("");
  useEffect(() => {
    (new FetchImageWorker()).handle(game.image).then(url => {
      setImage(url)
    }).catch(e => console.log(e));
  }, [game.image]);
  return (
    <>
      {image.length > 0 &&
      <div className="Game">
        <LazyLoad once={true} offset={100} height={150} placeholder={<Placeholder/>}>
          <div className="Game__container">
            {showNewRibbon(game)}
            <img data-testid="image" src={image} alt={game.name} onLoad={() => onLoad(image)}/>
            {showJackpot(jackpots.data, game)}
          </div>
          <div className="Game__play-button"></div>
        </LazyLoad>
      </div>
      }
    </>
  );
}

function onLoad(imageUrl: string) {
  URL.revokeObjectURL(imageUrl)
}

function showNewRibbon(game: GameInterface): string | JSX.Element {
  if(!game.categories.includes("new")) {
    return '';
  }
  return <Ribbon type="new" />;
}

function showJackpot(jackpotAggregate: JackpotsAggregate, game: GameInterface) {
  if(!jackpotAggregate.hasOwnProperty(game.id)) {
    return '';
  }
  return (
    <Jackpot amount={jackpotAggregate[game.id].amount} />
  );
}

const mapStateToProps = (state: StoreType) => ({
  jackpots: getJackpotsSelector(state)
})

export default connect(mapStateToProps)(Game);
