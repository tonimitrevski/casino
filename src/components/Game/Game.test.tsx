import React from 'react';
import { shallow } from 'enzyme';
import { Game } from "./Game";
import {JackpotsAggregate} from "../../core/aggregate/jackpots.aggregate";
import {JackpotStateInterface} from "../../redux/Jackpot/types/jackpot-state-interface";
import {GameInterface} from "../../core/models/game-interface";
import Ribbon from "./Ribbon/Ribbon";
import Jackpot from "./Jackpot/Jackpot";
const flushPromises = () => new Promise(setImmediate);
jest.mock("./../../core/services/fetch-image.worker.ts")
describe("Game Component", () => {
  let jackpotMock: JackpotStateInterface;
  let game: GameInterface;
  let imgUrl: string;
  beforeEach(() => {
    jackpotMock = {data: new JackpotsAggregate(), error: null};
    imgUrl = "https://casino-api-example.s3-eu-west-1.amazonaws.com/images/games/BSGHOULSGOLD.jpg";
    game = {
      categories: ['old'],
      id: 'test',
      image: imgUrl,
      name: 'Test Game',
    }
  });

  test('should test Game component', async () => {
    const wrapper = shallow(<Game jackpots={jackpotMock} game={game} />);
    wrapper.setProps({ image: imgUrl });
    setTimeout(() => {
      expect(wrapper.find('.Game__play-button').length).toBe(1);
      expect(wrapper.find('.Game__container').length).toBe(1);
      expect(wrapper.find("img").prop("src")).toEqual(imgUrl);
      expect(wrapper.find(Ribbon).length).toBe(0);
      expect(wrapper.find(Jackpot).length).toBe(0);
      expect(wrapper).toMatchSnapshot();
    }, 0);
  });

  test('Ribbon new is showing', async () => {
    game.categories = ["new"];
    const wrapper = shallow(<Game jackpots={jackpotMock} game={game} />);
    wrapper.setProps({ image: imgUrl });
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find(Ribbon).length).toBe(1);
    }, 0);
  });

  test('Jackpot is showing', async () => {
    jackpotMock.data["test"] = { game: "test", amount: 1000 };
    const wrapper = shallow(<Game jackpots={jackpotMock} game={game} />);
    wrapper.setProps({ image: imgUrl });
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find(Jackpot).length).toBe(1);
    }, 0);
  });
});



