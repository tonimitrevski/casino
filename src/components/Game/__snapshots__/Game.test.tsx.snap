// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Component Jackpot is showing 1`] = `
<DocumentFragment>
  <div
    class="Game"
  >
    <div
      class="lazyload-wrapper"
    >
      <div
        class="Game__container"
      >
        <img
          alt="Test Game"
          data-testid="image"
          src="https://casino-api-example.s3-eu-west-1.amazonaws.com/images/games/BSGHOULSGOLD.jpg"
        />
        <div
          class="jackpot"
        >
          <div
            class="jackpot__container"
          >
            <span
              class="amount"
            >
              Â£1,000.00
            </span>
          </div>
        </div>
      </div>
      <div
        class="Game__play-button"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Game Component Ribbon new is showing 1`] = `
<DocumentFragment>
  <div
    class="Game"
  >
    <div
      class="lazyload-wrapper"
    >
      <div
        class="Game__container"
      >
        <div
          class="ribbon"
        >
          <span
            class="ribbon__label"
          >
            new
          </span>
        </div>
        <img
          alt="Test Game"
          data-testid="image"
          src="https://casino-api-example.s3-eu-west-1.amazonaws.com/images/games/BSGHOULSGOLD.jpg"
        />
      </div>
      <div
        class="Game__play-button"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Game Component should test Game component 1`] = `
<DocumentFragment>
  <div
    class="Game"
  >
    <div
      class="lazyload-wrapper"
    >
      <div
        class="Game__container"
      >
        <img
          alt="Test Game"
          data-testid="image"
          src="https://casino-api-example.s3-eu-west-1.amazonaws.com/images/games/BSGHOULSGOLD.jpg"
        />
      </div>
      <div
        class="Game__play-button"
      />
    </div>
  </div>
</DocumentFragment>
`;
